torch/fx/experimental/symbolic_shapes.py

```diff
@@ -2885,14 +2983,39 @@ def get_shape_groups(self):
         return shape_groups
 
     @_lru_cache
-    def _maybe_evaluate_static(self, expr: "sympy.Expr", *, unbacked_only: bool = False) -> "Optional[sympy.Expr]":
+    def _maybe_evaluate_static(
+        self, expr: "sympy.Expr", *, unbacked_only: bool = False, compute_hint: bool = False
+    ) -> "Optional[sympy.Expr]":
         """
         Tries to evaluate expr without introducing guards
+
+        If unbacked_only == True, then we only do substitutions on
+        unbacked SymInts (leaving regular hinted integers alone).
         """
         expr = self.simplify(expr)
 
-        # Simplify making use of value range lower bound
         symbols = list(expr.free_symbols)
+
+        # Apply known runtime asserts
+        for s in symbols:
+            # Unbacked symints only
+            if s in self.var_to_val:
+                continue
+            subst = {}
+            for ra in self.deferred_runtime_asserts[s]:
+                if compute_hint:
+                    e = ra.expr.xreplace(self.var_to_val)
+                else:
+                    e = ra.expr
+                subst[e] = sympy.true
+                subst[sympy.Not(e)] = sympy.false
```

Comment:
do you just need a table of relations and flipping? How hard is this to add?

Reply:
It's not hard, but to be honest the use of subst to do these replacements is a hack in and of itself, so I didn't want to dig too far into the hole. Relevant Slack discuss about a better way to do this in sympy: https://pytorch.slack.com/archives/CFZ7V4DNZ/p1691504413431959